relay/pkg/block/forwarder.go

```diff
@@ -0,0 +1,110 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+	"github.com/keep-network/tbtc/relay/pkg/chain"
+)
+
+const (
+	// Size of the headers queue.
+	headersQueueSize = 50
+
+	// Maximum size of processed headers batch.
+	headersBatchSize = 5
+
+	// Maximum time for which the pulling process will wait for a single header
+	// to be delivered by the headers queue.
+	headerTimeout = 1 * time.Second
+
+	// Block duration of a Bitcoin difficulty epoch.
+	difficultyEpochDuration = 2016
+
+	// Duration for which the forwarder should rest after performing
+	// a push action.
+	forwarderPushingSleepTime = 45 * time.Second
+)
+
+var logger = log.Logger("relay-block-forwarder")
+
+// Forwarder takes blocks from the Bitcoin chain and forwards them to the
+// given host chain.
+type Forwarder struct {
+	btcChain  btc.Handle
+	hostChain chain.Handle
+
+	processedHeaders int
+
+	headersQueue chan *btc.Header
+	errChan      chan error
+}
+
+// RunForwarder creates an instance of the block forwarder and runs its
+// processing loop. The lifecycle of the forwarder loop can be managed
+// using the passed context.
+func RunForwarder(
+	ctx context.Context,
+	btcChain btc.Handle,
+	hostChain chain.Handle,
+) *Forwarder {
+	forwarder := &Forwarder{
+		btcChain:     btcChain,
+		hostChain:    hostChain,
+		headersQueue: make(chan *btc.Header, headersQueueSize),
+		errChan:      make(chan error, 1),
+	}
+
+	go forwarder.pushingLoop(ctx)
+
+	return forwarder
+}
+
+func (f *Forwarder) pushingLoop(ctx context.Context) {
+	logger.Infof("running forwarder pushing loop")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Infof("forwarder context is done")
+			return
+		default:
+			logger.Infof("pulling new headers from queue")
+
+			headers := f.pullHeadersFromQueue(ctx)
+			if len(headers) == 0 {
+				continue
+			}
+
+			logger.Infof(
+				"pushing [%v] header(s) to host chain",
+				len(headers),
+			)
+
+			if err := f.pushHeadersToHostChain(ctx, headers); err != nil {
+				f.errChan <- fmt.Errorf("could not push headers: [%v]", err)
+				return
```

Comment:
Let's stay that `pushHeadersToHostChain` failed. If we `return` and restart this loop or just log an error and `continue` without doing anything else, `headers` will not be pushed and we'll be waiting for another batch of headers from `pullHeadersFromQueue`. Is it blocking the maintainer on-chain? Can we afford to drop one batch of headers?

Reply:
It's even better to `return`, restart, and drop the batch entirely. In that case, the new forwarder instance will initialize the pull side properly basing on `Relay` contract state. If we just log and continue, we will receive the next batch from the queue which cannot be pushed on the host chain as the chain still "waits" for the old batch. The forwarder will remain in an inconsistent state.